File: python3/bron_kerbosch4.py
# coding: utf-8

from graph import UndirectedGraph, Vertex
from reporter import Reporter
from typing import Any, List, Optional, Set, Tuple


def explore(graph: UndirectedGraph, reporter: Reporter):
    '''Naive Bron-Kerbosch algorithm optimized'''
    candidates = graph.connected_nodes()
    if candidates:
        visit(
            graph=graph,
            reporter=reporter,
            candidates=candidates,
            excluded=set(),
            clique=None)


VertexStack = Optional[Tuple[Any, Vertex]]


def append_to(l: List[Vertex], clique: VertexStack):
    if clique is not None:
        append_to(l, clique[0])
        l.append(clique[1])


def visit(graph: UndirectedGraph, reporter: Reporter, candidates: Set[Vertex],
          excluded: Set[Vertex], clique: VertexStack):
    reporter.inc_count()
    if not candidates and not excluded:
        lst = []
        append_to(lst, clique)
        reporter.record(lst)

    while candidates:
        v = candidates.pop()
        neighbours = graph.adjacencies[v]
        assert neighbours
        visit(
            graph=graph,
            reporter=reporter,
            candidates=candidates.intersection(neighbours),
            excluded=excluded.intersection(neighbours),
            clique=(clique, v))
        excluded.add(v)

File: rust/bron_kerbosch/src/bron_kerbosch4.rs
//! Naive Bron-Kerbosch algorithm, optimized

use graph::{UndirectedGraph, Vertex};
use reporter::Reporter;
use util::intersect;
use vertex_stack::VertexStack;

use std::collections::HashSet;

pub fn explore(graph: &UndirectedGraph, reporter: &mut Reporter) {
    let candidates = graph.connected_nodes();
    if !candidates.is_empty() {
        visit(
            graph,
            reporter,
            candidates,
            HashSet::new(),
            VertexStack::Empty,
        );
    }
}

fn visit(
    graph: &UndirectedGraph,
    reporter: &mut Reporter,
    mut candidates: HashSet<Vertex>,
    mut excluded: HashSet<Vertex>,
    clique: VertexStack,
) {
    debug_assert!(candidates.iter().all(|&v| graph.degree(v) > 0));
    debug_assert!(excluded.iter().all(|&v| graph.degree(v) > 0));
    reporter.inc_count();
    if candidates.is_empty() && excluded.is_empty() {
        reporter.record(clique.collect());
        return;
    }

    while let Some(v) = remove_arbitrary(&mut candidates) {
        let neighbours = graph.adjacencies(v);
        debug_assert!(!neighbours.is_empty());
        let neighbouring_candidates = intersect(&neighbours, &candidates).cloned().collect();
        let neighbouring_excluded = intersect(neighbours, &excluded).cloned().collect();
        visit(
            graph,
            reporter,
            neighbouring_candidates,
            neighbouring_excluded,
            VertexStack::Cons(&clique, v),
        );
        excluded.insert(v);
    }
}

fn remove_arbitrary(s: &mut HashSet<Vertex>) -> Option<Vertex> {
    s.iter().next().cloned().map(|v| {
        s.remove(&v);
        v
    })
}
